server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -484,6 +485,9 @@ public boolean hasIndexAbstraction(String indexAbstraction, ClusterState state)
     public String resolveDateMathExpression(String dateExpression) {
         // The data math expression resolver doesn't rely on cluster state or indices options, because
         // it just resolves the date math to an actual date.
+        if (false == DateMathExpressionResolver.isDateMathExpression(dateExpression)) {
+            return dateExpression;
+        }
         return DateMathExpressionResolver.resolveExpression(dateExpression,
             new Context(null, null, getSystemIndexAccessLevel(), getSystemIndexAccessPredicate(), getNetNewSystemIndexPredicate()));
     }
```

Comment:
For ~10K indices, this change saves about 1ms by avoid instantiate the `Context` object. There are a few other method in this class that this same change can be applied. I didn't touch them to (1) keep the scope limited and (2) they are not related to the current problem.

Reply:
I see the `isDateMathExpression` escape condition is duplicated here from inside `DateMathExpressionResolver#resolveExpression`. I suppose the comp savings you mention come from not building the `Context` object.

This change sounds reasonable to me, although duplicating escape conditions on the execution path rings bells of fragility on my mind, but it this case this is introduced for performance reasons.

But, looking at how the `Context` object is actually used, I have an alternative suggestion.
The `DateMathExpressionResolver#resolveExpression` only needs the "current" date to do its job. Let's change the method interface to pass in the date directly. We can then remove the above comment that explains how date math doesn't rely on cluster state, etc (it will be clear with the new interface), and we also avoid constructing useless `Context` objects, with various null parameters, by the other callers of this method (everytime the caller builds such a context it must look at how it is going to be used).
What do you think?

Reply:
I was trying to scope the changes. But I think you have a point here. Instead of directly passing in the date, I changed the method to take a `LongSupplier` because reading system time is not necessarily cheap. Using `LongSupplier` means creating a Lambda each time, but initial testing suggest it does not incur obvious overhead (definitely still much better than creating `Context`). Thanks!